engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeHelpers.java

```diff
@@ -0,0 +1,161 @@
+  }
+
+  /** Checks whether an array like object is considered immutable.
+   * Immutable objects are instances of {@link EnsoObject} and can be safely cast
+   * to that interface.
+   *
+   * @param obj the object to check
+   * @return if the {@code obj} is already seen as immutable
+   */
+  public static boolean isImmutable(Object obj) {
+    return obj instanceof Vector;
+  }
+
+  /**
+   * Takes a slice from an array like object.
+   *
+   * @param self array like object
+   * @param start start of the slice
+   * @param end end of the slice
+   * @param len the length of the array
+   * @return an array-like object representing the slice
+   */
+  public static Object slice(Object self, long start, long end, long len) {
+    var slice = ArraySlice.createOrNull(self, start, len, end);
+    return slice == null ? self : slice;
+  }
+
+  /**
+   * Creates an uninitialized array of the given size.The values must be filled before the array is
+   * returned to Enso.
+   *
+   * @param size the size of the created array.
+   * @return the array instance
+   */
+  public static EnsoObject allocate(long size) {
+    var arr = new Object[Math.toIntExact(size)];
+    return new Array(arr);
+  }
+
+  @Builtin.Method(
+      name = "vector_from_function",
+      description = "Creates new Vector with given length and provided elements.",
+      autoRegister = false)
+  @Builtin.Specialize()
+  public static Object vectorFromFunction(
+      VirtualFrame frame,
+      long length,
+      Function fun,
+      State state,
+      @Cached("buildWithArity(1)") InvokeFunctionNode invokeFunctionNode,
+      @CachedLibrary(limit="3") WarningsLibrary warnings
+    ) {
+    var len = Math.toIntExact(length);
+    var target = ArrayBuilder.newBuilder(len);
+    boolean nonTrivialEnsoValue = false;
+    for (int i = 0; i < len; i++) {
+      var value = invokeFunctionNode.execute(fun, frame, state, new Long[] {(long) i});
+      if (value instanceof DataflowError) {
+        return value;
+      }
+      if (warnings.hasWarnings(value)) {
+        nonTrivialEnsoValue = true;
+      } else {
+        var isEnsoValue = value instanceof EnsoObject || value instanceof Long || value instanceof Double;
+        if (!isEnsoValue) {
+          nonTrivialEnsoValue = true;
+        }
+      }
+      target.add(value);
+    }
+    var res = target.toArray();
+    if (res instanceof long[] longs) {
+      return Vector.fromLongArray(longs);
+    }
+    if (res instanceof double[] doubles) {
+      return Vector.fromDoubleArray(doubles);
+    }
+    if (nonTrivialEnsoValue) {
+      return Vector.fromInteropArray(new Array((Object[])res));
+    } else {
+      return Vector.fromEnsoOnlyArray((Object[])res);
+    }
+  }
+
+  @Builtin.Method(
+      name = "vector_to_array",
+      description = "Returns an Array representation of this Vector.")
+  public static Object vectorToArray(Object obj) {
+    if (obj instanceof Vector.Generic vector) {
```

Comment:
I don't think it would make any performance difference. The generated node is:
```java
@BuiltinMethod(type = "Array_Like_Helpers", name = "vector_to_array", description = """
Returns an Array representation of this Vector.""")
@ImportStatic(org.enso.interpreter.runtime.data.vector.ArrayLikeHelpers.class)
public class VectorToArrayArrayLikeHelpersNode extends Node {

  java.lang.Object execute(java.lang.Object obj) {
    return ArrayLikeHelpers.vectorToArray(obj);
  }
}
```
The `vectorToArray` method will be inlined. Thus there will be one `instanceof` check and a `return`. `@Specialization` would also need to do `instanceof` checks...